frontend-api/src/main/java/uk/gov/di/authentication/frontendapi/lambda/LoginHandler.java

```diff
@@ -70,22 +79,64 @@ public APIGatewayProxyResponseEvent handleRequestWithUserContext(
         }
 
         try {
-            var nextState =
-                    stateMachine.transition(
-                            session.get().getState(), USER_ENTERED_VALID_CREDENTIALS, userContext);
             LoginRequest loginRequest = objectMapper.readValue(input.getBody(), LoginRequest.class);
             boolean userHasAccount = authenticationService.userExists(loginRequest.getEmail());
             if (!userHasAccount) {
                 LOGGER.error("The user does not have an account");
                 return generateApiGatewayProxyErrorResponse(400, ErrorResponse.ERROR_1010);
             }
+
+            SessionState currentState = session.get().getState();
+            boolean keyExists =
+                    codeStorageService.hasEnteredPasswordIncorrectBefore(loginRequest.getEmail());
+
+            if (!keyExists && currentState.equals(ACCOUNT_TEMPORARILY_LOCKED)) {
+                var nextState =
+                        stateMachine.transition(
+                                session.get().getState(), ACCOUNT_LOCK_EXPIRED, userContext);
+                sessionService.save(session.get().setState(nextState));
+            }
+
             boolean hasValidCredentials =
                     authenticationService.login(
                             loginRequest.getEmail(), loginRequest.getPassword());
+
+            boolean hasEnteredPasswordIncorrectBefore =
+                    codeStorageService.hasEnteredPasswordIncorrectBefore(loginRequest.getEmail());
+
             if (!hasValidCredentials) {
+                if (hasEnteredPasswordIncorrectBefore) {
+                    int count =
+                            codeStorageService.getIncorrectPasswordCount(loginRequest.getEmail());
+
+                    if (count >= MAX_INCORRECT_PASSWORD_ATTEMPTS) {
```

Comment:
Should we increase the count before this check?

Reply:
I have implemented this such that once we exceed the attempt limit then we stop increasing the recorded count value in redis as that would reset the expiry for the account lock back to 15 mins. I could add more code to keep track of the failed passwords past the limit without reseting the expiry in redis but that seems redundant in terms of what we are trying to achieve. Im not too precious either way but lets see what the others say.

Reply:
Ok that makes sense 👍